src/BloomBrowserUI/bookEdit/bloomField/bloomField.ts

```diff
@@ -0,0 +1,254 @@
+            sel.removeAllRanges();
+            sel.addRange(range);
+        } else {
+            var p = $(field).find('p')[0];
+            if (!p) {
+                return; // these have text, but not p's yet. We'll have to wait until they leave (blur) to add in the P's.
+            }
+            var range = document.createRange();
+            range.selectNodeContents(p);
+            range.collapse(true); //move to start of first paragraph
+
+            var sel = window.getSelection();
+            sel.removeAllRanges();
+            sel.addRange(range);
+        }
+    }
+
+
+    static MoveCursorToFirstParagraph(field: HTMLElement, position: CursorPosition ){
+        var range = document.createRange();
+        if(position === CursorPosition.start) {
+            range.selectNodeContents($(field).find('p').first()[0]);
+        }
+        else{
+            range.selectNodeContents($(field).find('p').last()[0]);
+        }
+        range.collapse(position === CursorPosition.start);//true puts it at the start
+        var sel = window.getSelection();
+        sel.removeAllRanges();
+        sel.addRange(range);
+    }
+    static ManageWhatHappensIfTheyDeleteEverything(field: HTMLElement) {
+        // if the user types (ctrl+a, del) then we get an empty element or '<br></br>', and need to get a <p> in there.
+        // if the user types (ctrl+a, 'blah'), then we get
+
+        $(field).on("input", function (e) {
+            if ($(this).find('p').length === 0) {
+                BloomField.ModifyForParagraphMode(this);
+
+                // Now put the cursor in the paragraph, *after* the character they may have just typed or the
+                // text they just pasted.
+                BloomField.MoveCursorToFirstParagraph(field, CursorPosition.end);
+            }
+        });
+    }
+    // Some custom templates have image containers embedded in bloom-editable divs, so that the text can wrap
+    // around the picture. The problems is that the user can do (ctrl+a, del) to start over on the text, and
+    // inadvertently remove the embedded images. So we introduced the "bloom-preventRemoval" class, and this
+    // tries to safeguard element bearing that class.
+
+    static PreventRemovalOfSomeElements(field:HTMLElement) {
+        var numberThatShouldBeThere = $(field).find(".bloom-preventRemoval").length;
+        if(numberThatShouldBeThere > 0) {
+            $(field).on("input", function (e) {
+                if ($(this).find(".bloom-preventRemoval").length < numberThatShouldBeThere) {
+                    document.execCommand('undo');
+                    e.preventDefault();
+                }
+            });
+        }
+
+        //OK, now what if the above fails in some scenario? This adds a last-resort way of getting
+        //bloom-editable back to the state it was in when the page was first created, by having
+        //the user type in RESETRESET and then clicking out of the field.
```

Comment:
Maybe add something like "Since the elements that should not be deleted are part of a parallel field in a template language, initial page setup will copy it into a new version of the messed up one if the relevant language version is missing altogether"